src/main.rs

```diff
@@ -1,69 +1,41 @@
 // SPDX-FileCopyrightText: 2025 blinry <mail@blinry.org>
 //
 // SPDX-License-Identifier: GPL-3.0-or-later
+mod process_tree;
 
-use std::path::PathBuf;
+use std::{
+    io::{stdout, Write},
+    path::PathBuf,
+};
 
-#[derive(Debug)]
-struct Process {
-    cwd: PathBuf,
-    tty: i32,
-}
-
-#[derive(Debug)]
-struct ProcessTree {
-    node: Process,
-    children: Vec<ProcessTree>,
-}
-
-impl ProcessTree {
-    fn new(pid: u32) -> ProcessTree {
-        let p = procfs::process::Process::new(pid as i32).expect("Process should exist");
-        let node = Process {
-            cwd: p.cwd().expect("Process should have a cwd"),
-            tty: p.stat().expect("Process should hav status info").tty_nr,
-        };
-
-        let t = p
-            .task_main_thread()
-            .expect("Process should have main thread");
-        let c_pids = t.children().expect("Task should have children");
-        let children = c_pids
-            .iter()
-            .map(|&c_pid| ProcessTree::new(c_pid))
-            .collect();
-        ProcessTree { node, children }
-    }
-
-    fn leaf_nodes_with_tty(&self) -> Vec<&Process> {
-        let leaves: Vec<&Process> = self
-            .children
-            .iter()
-            .flat_map(|c| c.leaf_nodes_with_tty())
-            .collect();
-
-        if !leaves.is_empty() {
-            return leaves;
-        }
-
-        if self.node.tty != 0 {
-            vec![&self.node]
-        } else {
-            vec![]
-        }
-    }
-}
+use process_tree::ProcessTree;
+use procfs::ProcResult;
 
-fn main() {
+fn get_path() -> ProcResult<PathBuf> {
     let t = ProcessTree::new(
         std::env::args()
             .nth(1)
-            .expect("PID should be provided as first argument")
-            .parse()
-            .expect("PID should be a number"),
-    );
-    let leaves = t.leaf_nodes_with_tty();
+            .ok_or("First argument is required.")?
+            .parse()?,
+    )?;
 
-    // Print cwd of first leaf.
-    println!("{}", leaves[0].cwd.display());
+    t.into_deepest_leaf().map(Into::into)
+}
+
+fn main() -> std::io::Result<()> {
+    let path = match get_path() {
+        Ok(path) => path.into(),
+        Err(error) => {
+            eprintln!("Could not get cwd: {error}");
+            if let Some(home) = std::env::var_os("HOME") {
+                home
+            } else {
+                eprintln!("HOME not set");
+                "/".into()
+            }
+        }
+    };
```

Comment:
It would be kind of neat to be able to express this in a less nested way. In my head, it's something like "here are three options, use the first one that can be computed, and use different error messages if they can't be used".

If you'd enjoy refactoring this a bit in this way, that'd make me happy – but we can also do it later!